src/gf.c

```diff
@@ -1621,56 +1621,59 @@ void JL_NORETURN jl_method_error_bare(jl_function_t *f, jl_value_t *args, size_t
 
 void JL_NORETURN jl_method_error(jl_function_t *f, jl_value_t **args, size_t na, size_t world)
 {
-    jl_value_t *argtup = jl_f_tuple(NULL, args+1, na-1);
+    jl_value_t *argtup = jl_f_tuple(NULL, args, na - 1);
     JL_GC_PUSH1(&argtup);
     jl_method_error_bare(f, argtup, world);
     // not reached
 }
 
-jl_tupletype_t *arg_type_tuple(jl_value_t **args, size_t nargs)
+jl_tupletype_t *arg_type_tuple(jl_value_t *arg1, jl_value_t **args, size_t nargs)
 {
     jl_tupletype_t *tt;
     size_t i;
-    if (nargs * sizeof(jl_value_t*) < jl_page_size) {
-        jl_value_t **types;
-        JL_GC_PUSHARGS(types, nargs);
-        for (i = 0; i < nargs; i++) {
-            jl_value_t *ai = args[i];
-            if (jl_is_type(ai))
-                types[i] = (jl_value_t*)jl_wrap_Type(ai);
-            else
-                types[i] = jl_typeof(ai);
-        }
-        // if `ai` has free type vars this will not be a valid (concrete) type.
-        // TODO: it would be really nice to only dispatch and cache those as
-        // `jl_typeof(ai)`, but that will require some redesign of the caching
-        // logic.
-        tt = jl_apply_tuple_type_v(types, nargs);
-        JL_GC_POP();
+    int onstack = (nargs * sizeof(jl_value_t*) < jl_page_size);
+    jl_value_t **roots;
+    jl_value_t **types;
+    JL_GC_PUSHARGS(roots, onstack ? nargs : 1);
+    if (onstack) {
+        types = roots;
     }
     else {
-        jl_svec_t *types = jl_alloc_svec(nargs);
-        JL_GC_PUSH1(&types);
-        for (i = 0; i < nargs; i++) {
-            jl_value_t *ai = args[i];
-            if (jl_is_type(ai))
-                jl_svecset(types, i, (jl_value_t*)jl_wrap_Type(ai));
-            else
-                jl_svecset(types, i, jl_typeof(ai));
+        roots[0] = (jl_value_t*)jl_alloc_svec(nargs);
+        types = jl_svec_data(roots[0]);
+    }
+    for (i = 0; i < nargs; i++) {
+        jl_value_t *ai = (i == 0 ? arg1 : args[i - 1]);
+        if (jl_is_type(ai)) {
+            // if `ai` has free type vars this will not be a valid (concrete) type.
+            // TODO: it would be really nice to only dispatch and cache those as
+            // `jl_typeof(ai)`, but that will require some redesign of the caching
+            // logic.
+            types[i] = (jl_value_t*)jl_wrap_Type(ai);
+            if (!onstack)
+                jl_gc_wb(roots[0], types[i]);
+        }
+        else {
+            types[i] = jl_typeof(ai);
         }
-        tt = jl_apply_tuple_type(types);
-        JL_GC_POP();
     }
+    if (onstack)
+        tt = jl_apply_tuple_type_v(types, nargs);
+    else
+        tt = jl_apply_tuple_type((jl_svec_t*)roots[0]);
+    JL_GC_POP();
     return tt;
 }
 
 jl_method_instance_t *jl_method_lookup(jl_methtable_t *mt, jl_value_t **args, size_t nargs, int cache, size_t world)
 {
-    jl_typemap_entry_t *entry = jl_typemap_assoc_exact(mt->cache, args, nargs, jl_cachearg_offset(mt), world);
+    if (nargs == 0)
```

Comment:
Could this be an assertion?

Reply:
Yes, it looks like the callers have that precondition (since they were computing `mt`)